lib/core/nuxt.js

```diff
@@ -105,30 +105,23 @@ export default class Nuxt {
     }
   }
 
-  getObjectHooks(hooksObj, keyList = [], stack = '') {
-    Object.keys(hooksObj).forEach((key) => {
-      if (typeof hooksObj[key] === 'object' && hooksObj[key] !== null) {
-        this.getObjectHooks(hooksObj[key], keyList, `${stack}:${key}`)
+  flatHooks(configHooks, hooks = {}, parentName) {
+    Object.keys(configHooks).forEach((key) => {
+      const subHook = configHooks[key]
+      const name = parentName ? `${parentName}:${key}` : key
+      if (typeof subHook === 'object' && subHook !== null) {
+        this.flatHooks(subHook, hooks, name)
       } else {
-        const value = hooksObj[key]
-        key = `${stack}:${key}`
-        keyList.push({
-          name: key.slice(1),
-          hook: value
-        })
+        hooks[name] = subHook
       }
     })
-    return keyList.reduce((hash, pair) => {
-      return { ...hash, [pair.name]: pair.hook }
-    }, {})
+    return hooks
   }
 
-  addObjectHooks(obj) {
-    const hooksObj = this.getObjectHooks(obj)
-    Object.keys(hooksObj).filter(Boolean).forEach((key) => {
-      const hook = hooksObj[key]
-      const hooks = Array.isArray(hook) ? hook : [hook]
-      hooks.forEach(h => this.hook(key, h))
+  addHooks(configHooks) {
+    const hooks = this.flatHooks(configHooks)
+    Object.keys(hooks).filter(Boolean).forEach((key) => {
+      [].concat(hooks[key]).forEach(h => this.hook(key, h))
```

Comment:
@clarkdo can you use `Array.prototype.concat` here?

Reply:
Actually cannot, if `hooks[key]` is not array, it cannot convert.

Reply:
`[].concat('test')` works.
`Array.prototype.concat.apply([], 'tset')` doesn't work.

Reply:
@clarkdo oh gotcha

Reply:
@clarkdo erm, actually I think you'd need `.call()`, not `.apply()` -- but it's all good :)

Reply:
Yes, call works. But it makes no difference in performance and memory. 😸

Reply:
And also if `hooks` amount is huge (won't happen in real case), `[].concat` should be faster than `Array.prototype.concat.call`, because former one doesn't resolve scope as many times as the latter one.